### DISCUSSION TITLE
Support automated space insertion with the Pangu library.

### DISCUSSION THREAD

**Discussion by cmpute** [yuin/goldmark#351] [Ideas]
Pangu is a module that automatically insert spaces between CJK characters and Latin characters, which is very useful to increase readability of some text mixed with CJK and Latin characters. It has been implemented in several languages including [Go](https://github.com/vinta/pangu) and [JS](https://github.com/vinta/pangu.js).

I posted a feature request for this in the Hugo repo (https://github.com/gohugoio/hugo/issues/10617), and I agreed on that this should be implemented in the renderer side. Therefore, I propose to add an CJK extension to allow automatic space insertion with the Pangu library.

**Comment by yuin** [comment]
README says "Depends only on standard libraries."
So this is unlikely to happen.

**Reply by cmpute** [reply]
The [Go implementation](https://github.com/vinta/pangu/blob/master/pangu.go) is self-contained (without third-party dependencies), and it has only 147 lines. Do you consider directly porting the source code in? The licenses are also compatible (both MIT).

**Reply by yuin** [reply]
It seams that the implementation has problems:

- It seems using a lot of `regexp`. Go's regexp performance is poor for common usecases(but stable for text length). goldmark avoids `regexp` as much as possible.
- Markdown processing is not equal to simple text processing. Parser parses markdown to AST, then text are spitted into peace of segments.
  - ie: `こんにちは[**Tomさん**](http://example.com/tom)ごきげんいかがですか` -> `[Text(こんにちわ),Link(label=Empasis(Text(Tomさん)), url="http://example.com/tom"), Text(ごきげんいかがですか)]`
  - So simple regexp approach will not be worked for this kind of structure.

**Reply by cmpute** [reply]
Ok thanks for the explanation. I still wish this feature could be supported, but it seems like there is an amount of required work to implement the spacing feature. Again, thanks for your great work on this project!

**Reply by cmpute** [reply]
Is regexp a deal breaker for this functionality? If it's acceptable when used only in this functionality, I do want to have a try to implement this feature.

**Reply by yuin** [reply]
First of all, we should consider whether this functionalities are possible to implement as an extension. 
If so, implementation details are not an issue.

**Reply by cmpute** [reply]
I have an idea of how to implement this. First, we can define it as an extension just like the CJK extension (or embed it into the CJK extension as an option), then we process it just like the `EastAsianLineBreaksCSS3Draft` option, by modifying the html rendering part:

https://github.com/yuin/goldmark/blob/6951ff3bd1825c85dc9a0197ff9b33825588c26f/renderer/html/html.go#L652-L687

Since this functionality is mainly for html rendering, so I think it's okay to insert the space adding logics into here.

**Reply by cmpute** [reply]
The modification can also happends during AST generation, but in this case it will affect all renderers, and maybe unable to be packed as an extension. What's your opinion on this please? @yuin 

I'm willing to contribute if we can agree on a good solution strategy

**Reply by axmmisaka** [reply]
I was looking into getting pangu.js applied in hugo and came upon this issue; after some code reading and vibe code reading I think we definitely are able to integrate pangu's logic at AST generation stage (i.e. walk the AST after it's created, and apply pangu vertex by vertex; some special edge cases might need to be handled though).

Pangu.js did the transformation on HTML so functionally it's 100% doable in goldmark, but they have JS DOM API in browsers; we don't have that in go. iiuc Pangu.go is simple because it only work on plaintext.

Affecting all renderers will not, imo, impact the ability to pack it as an extension, as it could be turned on/off. 

Please do correct me if I'm wrong, I'm a long time user but first time contributor.
